src/persist-txn/src/txn_read.rs

```diff
@@ -0,0 +1,288 @@
+
+    next_batch_id: usize,
+    /// The batches needing application as of the current progress.
+    ///
+    /// This is indexed by a "batch id" that is internal to this object because
+    /// timestamps are not unique.
+    ///
+    /// Invariant: Values are sorted by timestamp.
+    unapplied_batches: BTreeMap<usize, (ShardId, String, u64)>,
+    /// Invariant: An index into `unapplied_batches` keyed by the serialized
+    /// batch.
+    batch_idx: HashMap<String, usize>,
+    /// The times at which each data shard has been written.
+    ///
+    /// Invariant: Times in the Vec are in ascending order.
+    pub(crate) datas: BTreeMap<ShardId, Vec<u64>>,
+}
+
+impl TxnsCache {
+    pub(crate) async fn init(
+        init_ts: u64,
+        txns_read: ReadHandle<ShardId, String, u64, i64>,
+        txns_write: &mut WriteHandle<ShardId, String, u64, i64>,
+    ) -> Self {
+        let () = crate::empty_caa(|| "txns init", txns_write, init_ts).await;
+        Self::open(init_ts, txns_read).await
+    }
+
+    async fn open(init_ts: u64, txns_read: ReadHandle<ShardId, String, u64, i64>) -> Self {
+        // TODO(txn): Figure out the compaction story. This might require
+        // sorting inserts before retractions within each timestamp.
+        let subscribe = txns_read
+            .subscribe(Antichain::from_elem(0))
+            .await
+            .expect("handle holds a capability");
+        let mut ret = TxnsCache {
+            progress_exclusive: 0,
+            txns_subscribe: subscribe,
+            next_batch_id: 0,
+            unapplied_batches: BTreeMap::new(),
+            batch_idx: HashMap::new(),
+            datas: BTreeMap::new(),
+        };
+        ret.update_gt(init_ts).await;
+        ret
+    }
+
+    /// Returns the minimum timestamp at which the data shard can be read.
+    ///
+    /// Returns Err if that data has not been registered.
+    pub fn data_since(&self, data_id: &ShardId) -> Result<u64, NotRegistered> {
+        let time = self
+            .datas
+            .get(data_id)
+            .and_then(|times| times.first().copied());
+        time.ok_or(NotRegistered {
+            data_id: *data_id,
+            ts: self.progress_exclusive,
+        })
+    }
+
+    /// Returns the greatest timestamp written to the data shard that is `<=`
+    /// the given timestamp.
+    ///
+    /// A data shard might be definite at times past the physical upper because
+    /// of invariants maintained by this txn system. As a result, this method
+    /// "translates" an `as_of` of a persist snapshot into the one that will
+    /// resolve on the physical shard.
+    ///
+    /// WIP I feel like the above could be explained better.
+    ///
+    /// If necessary, blocks until the txn shard upper passes the requested
+    /// timestamp (closing it out to future writes). Will not block if a call to
+    /// `update_past` (or a previous call to this method) with the same input
+    /// has resolved.
```

Comment:
I might be missing something but I don't see where the blocking is happening. It looks like the method will instead panic on `assert!(self.progress_exclusive > ts);` if the the txn shard upper (`progress_exclusive`)  is not passed the requested timestamp (`ts`).

Is it up to the caller to call `update_ge(ts)` before calling this method? If so, should we just move that call into this method?

Reply:
oh you're right, this is the panic case. updated

my thinking was to have `update_gt` as a manual precondition because I think it'd be nice for this method to be obviously non-blocking (by not being an async fn). definitely willing to revisit this if we think it ends up being more of a footgun than it helps